### PR TITLE
PM-196: finish gno reward address claim banner and scoreboard element

### PR DESCRIPTION
**Pull request by andre-meyer**
Screenshots:
image
image
(Not my Address so I censored it)

  
image

### CODE REVIEW

File: src/components/HeaderRewardStatus/index.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      return null
+    }
+
+    return (
+      <div className="headerReward">
+        <div className="container">
+          <div className="col-xs-12 headerRewardStatus">
+            <i className="icon icon--warning" />
+
+            <span>No reward claim address specified yet. <Link className="headerReward__link" to="/scoreboard" href="/scoreboard" onClick={openSetMainnetAddressModal}>Setup claim address</Link></span>
```
Comment (mmvsha) [inline src/components/HeaderRewardStatus/index.js]: Can it be written in multiple lines and not only one?

File: src/components/HeaderRewardStatus/index.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+class HeaderRewardStatus extends Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      fetchingMainnetAddress: true,
+    }
+  }
+  async componentWillMount() {
+    await this.props.requestMainnetAddress()
```
Comment (mmvsha) [inline src/components/HeaderRewardStatus/index.js]: does this actually prevent component from rendering? If no, I think there is no point in using `async/await`

**Comment by andre-meyer** [comment]
@apanizo 
Can you elaborate where you mean? `requestMainnetAddress` action in `actions/blockchain` is used for the HeaderRewardStatus and the Scoreboard. `updateMainnetAddress` is used in Modal. I did not create any reducers

File: src/api/gnosis.js [hunk 3]
```diff
@@ -31,11 +33,29 @@ export const initGnosisConnection = async (GNOSIS_OPTIONS) => {
...
+      resolve(gnosisInstance)
+    }, 50)
+
+    setTimeout(() => {
+      if (stillRunning) {
+        clearInterval(instanceCheck)
+        reject(new Error('Connection to Gnosis.js timed out'))
+      }
+    }, 15000)
+  })
```
Comment (andre-meyer) [inline src/api/gnosis.js]: This unrelated change allows actions using `getGnosisConnection` to wait until Gnosis is running. This is a sort of race-condition that could happen in really rare conditions

**Comment by andre-meyer** [comment]
@apanizo Added the screenshot for the Header Element

File: src/api/gnosis.js [hunk 4]
```diff
@@ -31,11 +33,29 @@ export const initGnosisConnection = async (GNOSIS_OPTIONS) => {
  * Returns an instance of the connection to GnosisJS
  */
 export const getGnosisConnection = async () => {
-  if (!gnosisInstance) {
-    throw new Error('GnosisJS not initialized yet')
+  if (gnosisInstance) {
```
Comment (apanizo) [inline src/api/gnosis.js]: I understand the code and the reasons because you have done this. But I think this hides a deeper problem. AFAIK in one part of the code, we initialize the gnosis connection, and in other, we try to access it. That is a race condition and a very big potential problem. 

Solving it with intervals is not a proper solution. It is a hack :)

After some time thinking about it you can solve it using Proxies or one JS library like ```Watch.JS``` but I think we should go one level up.

So, I suggest two options, load the application once we assured the gnosisConnection is set, which I am very against. Two try to investigate where we call that function before it is initialized. 

I bet option 2. As far as I see we call ```getGnosisConnection``` in this file and also in ```actions/market.js```, to my understanding until the connection is loaded one user cannot do ANYTHING on the website, and also exported methods from this file are also only used in actions, right? 

So, which could be a scenario where we can have this problem?

File: src/api/gnosis.js [hunk 5]
```diff
@@ -438,3 +458,18 @@ export const getOlympiaTokensByAccount = async (account) => {
   const balance = await gnosis.olympiaToken.balanceOf(account)
   return new Decimal(balance.toFixed(0)).toString()
 }
+
+export const getMainnetAddressForRinkebyAccount = async (account) => {
```
Comment (apanizo) [inline src/api/gnosis.js]: Do not think it is a good idea to have 400 LOC file. After seeing this file maybe it is a correct moment of split it in three parts:
```
api/gnosis.js <-- import and export functions from the next files
api/blockchain/trading.js
api/blockchain/gas.js
api/blockchain/network.js // or Olympia     
```

File: src/components/HeaderRewardStatus/index.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+class HeaderRewardStatus extends Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      fetchingMainnetAddress: true,
+    }
+  }
+  componentWillMount() {
```
Comment (apanizo) [inline src/components/HeaderRewardStatus/index.js]: Why you have used ```componentWillMount```? normally this method it is only used for not common configuration in root components. 
Instead of that, I suggest ```componentDidMount```, and react rerendering again the component once the property is available. 

In that way you have taking advatnage of the removed concept of time in React, simplified the state (probably you just need a functional component) and removed...

File: src/components/HeaderRewardStatus/index.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    this.setState({
+      fetchingMainnetAddress: false,
+    })
+  }
+
+  render() {
+    const { hasWallet, mainnetAddress, openSetMainnetAddressModal } = this.props
+    const { fetchingMainnetAddress } = this.state
+
+    const hideRewardStatusInfo = !hasWallet || mainnetAddress || fetchingMainnetAddress
```
Comment (apanizo) [inline src/components/HeaderRewardStatus/index.js]: ```fetchingMainnetAddress```

File: src/components/HeaderRewardStatus/index.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (hideRewardStatusInfo) {
+      return null
+    }
+
+    return (
+      <div className="headerReward">
```
Comment (apanizo) [inline src/components/HeaderRewardStatus/index.js]: No more Layout components and css-in-modules despite of this is a custom component focused on olympia?

Maybe we should find another location for those components? Something like ```components/olympia```?

File: src/components/SetMainnetAddress/index.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import InteractionButton from 'containers/InteractionButton'
+
+import './style.less'
+
+const FORM = 'setMainnetAddressForm'
+
+const SetMainnetAddress = ({
+  submitting, handleSubmit, error, submitFailed, closeModal,
+}) => (
+  <div className="setMainnetAddress">
```
Comment (apanizo) [inline src/components/SetMainnetAddress/index.js]: Same here, I think if you just use Layout components you will probably need less css.

File: src/components/SetMainnetAddress/index.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    <div className="setMainnetAddressModal">
+      <a className="setMainnetAddressModal__close" href="javascript:void(0);" onClick={() => closeModal()} />
```
Comment (apanizo) [inline src/components/SetMainnetAddress/index.js]: No functions inside the render method. Use a constant, otherwise, this component will be always rerendered.

File: src/components/SetMainnetAddress/index.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        {error && <p className="setMainnetAddressModal__error">{error}</p>}
+        {submitFailed && <p className="setMainnetAddressModal__error">Sorry, the transaction failed. Please try again later or contact us!</p>}
+      </form>
+    </div>
+  </div>
+)
+
+export default reduxForm({
+  form: FORM,
+  onSubmit: async (values, dispatch) => {
```
Comment (apanizo) [inline src/components/SetMainnetAddress/index.js]: Same here, extract this function as a constant.

File: src/routes/scoreboard/components/RewardClaim.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+AddressSection.defaultProps = {
+  mainnetAddress: '',
+}
+
+const NoAddressSection = ({ openSetMainnetAddressModal }) => (
+  <Block className={cx('noAddressBlock')}>
+    <Paragraph>
+      No address specified yet.
+    </Paragraph>
+    <Link to="/scoreboard" href="/scoreboard" onClick={openSetMainnetAddressModal}>Setup claim address</Link>
```
Comment (apanizo) [inline src/routes/scoreboard/components/RewardClaim.js]: ```to``` and ```href```? Is that necessary?

File: src/selectors/blockchain.js [hunk 13]
```diff
@@ -88,6 +88,16 @@ export const getCurrentNetworkId = (state) => {
   return undefined
 }
 
+export const getMainnetAddress = (state) => {
```
Comment (apanizo) [inline src/selectors/blockchain.js]: This selector I would suggest to move it to ```scr/routes/scoreboard/store/selectors``` (same reasons as for the actions)

File: src/routes/scoreboard/components/RewardClaim.js (hunk 12 above)
Comment (andre-meyer) [reply src/routes/scoreboard/components/RewardClaim.js]: eslint complains otherwise

File: src/components/SetMainnetAddress/index.js (hunk 11 above)
Comment (andre-meyer) [reply src/components/SetMainnetAddress/index.js]: But this isn't inside a `render` function

File: src/components/HeaderRewardStatus/index.js (hunk 7 above)
Comment (andre-meyer) [reply src/components/HeaderRewardStatus/index.js]: Elaborate?

File: src/components/SetMainnetAddress/index.js (hunk 11 above)
Comment (apanizo) [reply src/components/SetMainnetAddress/index.js]: I mean, you are passing a new object every time so in the internal checks, it thinks it has changed every  :)   { name: 'hello' } !== { name: 'hello'}

File: src/routes/scoreboard/components/RewardClaim.js (hunk 12 above)
Comment (apanizo) [reply src/routes/scoreboard/components/RewardClaim.js]: ok

File: src/components/HeaderRewardStatus/index.js (hunk 7 above)
Comment (apanizo) [reply src/components/HeaderRewardStatus/index.js]: I mean, if you implement changes suggested in the previous comment you do not need ```fetchingMainnetAddress```

File: src/components/HeaderRewardStatus/index.js (hunk 6 above)
Comment (andre-meyer) [reply src/components/HeaderRewardStatus/index.js]: I removed the state and changed this to look like this:
```
async componentDidMount() {
    await this.props.requestMainnetAddress()
  }

  render() {
    const { hasWallet, mainnetAddress, openSetMainnetAddressModal } = this.props

    const hideRewardStatusInfo = !hasWallet || mainnetAddress

    if (hideRewardStatusInfo) {
      return null
    }
...
```

Now the HeaderBar shows for a brief moment for users who have added their mainnetAddress already, how can I solve this?

`mainnetAddress` is undefined while it's loading and if no mainnetAddress wasn't defined.
See this video for an example: https://i.imgur.com/WiSdvdZ.gifv

Comment (apanizo) [reply src/components/HeaderRewardStatus/index.js]: Some suggestion to improve the UX:
*  Use localStorage for storing the info
*  The message at top of the window I would layout it as a ```position: fixed``` so when it appear or disappear the rest of the layout won't move.
*  Play with the default behaviour. If I am right, right now the default behaviour is "the user has NOT introduced their address", what happens if you revert it?

If you mix these three concepts, I think it should be fine.

Comment (andre-meyer) [reply src/components/HeaderRewardStatus/index.js]: > - Use localStorage for storing the info

I can add it to the localStorage Saving Middleware, but that wouldn't eliminate the problem, it would still happen on the first load, and when the user logs in with a different account, I would need to manually clear it. What do you think?
If I save it separately I can do:
```
{
  gnosisRewardAddresses: {
    0x123: 0x456,
    0x987: 0x654,
  }
}
```
but this seems like a lot of changes. Should I do it anyway?
> - The message at top of the window I would layout it as a position: fixed so when it appear or disappear the rest of the layout won't move.

This would overlay the Header/Nav Area. I could move the Header/Nav Area down, but then the site looks weird. Should I ask a Designer about this?
> - Play with the default behaviour. If I am right, right now the default behaviour is "the user has NOT introduced their address", what happens if you revert it?

Not sure If I understand this correctly. How can I change the default to "the user has entered their address"?

Comment (apanizo) [reply src/components/HeaderRewardStatus/index.js]: The local storage solution is something like: ```ALREADY_REGISTERED: true/false```, it is not a problem to switch user accounts because localSotrage is cleared when the user logs out.
Information stored in localStorage, should be ready before mounting in most of the cases. In that way, once the AJAX call is resolved and the store is updated, probably the component won't be rerendered. 

About the second point... nevermind. I think we can come back to it if the others do not work.

Third point. In the video you sent me, you show the message of "Remember to register your address..." by default, and then once the AJAX call is finished and the store is updated, you rerender the component having this glitch, right? What about if we consider the default behaviour is "the user has already registered" his address and just in case AJAX call says the opposite update the component? The effect would be better I think. 

Does it make sense?

Comment (andre-meyer) [reply src/components/HeaderRewardStatus/index.js]: Okay, I agree on all points.

But I don't think we can easily invert that behaviour. The Component has no concept of an AJAX Call, it simply receives a value from the store via a selector. I can add a variable to the provider indicating if the user has already registered their account or not - if that value is not available (`undefined`) , I won't display the Header, if the user has not registered it will be`false`, if they have registered it, it will be `true`. This way I have the information in the store, and I don't have to rely on an Ajax call in the Component.

Comment (apanizo) [reply src/components/HeaderRewardStatus/index.js]: When I say AJAX call I mean: ```this.props.requestMainnetAddress()``` does it clarify a bit my explanation?
What you say is exactly what I mean, why do you need to introduce the variable in the redux store -> provider?
 I think that would add more complexity. 

I suggest:
```
 
   const { addressRegistered } = this.props  //use this prop for knowing if user has already registered or not.
         //Where:
         //    undefined --> do nothing  on the top of the window. Show you have to register component in scoreboard
         //    false --> Prop updated from localstore or requestMainnetAddress. Show msg on top of the window and component in scoreboard with the warning
        //     true --> Prop updated from localstorage or requestMainnetAddress. Remove msg of the top window and the component in scoreboard saying all is good.
```

Does it make more sense now?

File: src/api/gnosis.js (hunk 5 above)
Comment (andre-meyer) [reply src/api/gnosis.js]: Not enough time for this, sorry

File: src/routes/scoreboard/store/actions/mainnetRegistry.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+export const requestMainnetAddress = () => async (dispatch, getState) => {
+  const state = getState()
+  const { name: providerName, ...provider } = getSelectedProvider(state)
+  const mainnetAddress = await getMainnetAddressForRinkebyAccount(provider.account)
+
+  await dispatch(updateProvider({
+    provider: providerName,
+    ...provider,
+    mainnetAddressAvailable: parseInt(mainnetAddress, 16) > 0, // returns 0x0... if not set
```
Comment (apanizo) [inline src/routes/scoreboard/store/actions/mainnetRegistry.js]: Is ```mainnetAddressAvailable```  property necessary? I think that data could easily be obtained from a selector, right? In that way, the store would be simpler, right?

File: src/routes/scoreboard/store/selectors/index.js [hunk 15]
```diff
@@ -34,6 +34,16 @@ export const firstOlympiaUsersSelectorAsList = createSelector(
     : undefined),
 )
 
+export const olympiaMainnetRegistryAddress = (state) => {
+  const provider = getSelectedProvider(state)
+  
+  if (provider) {
```
Comment (apanizo) [inline src/routes/scoreboard/store/selectors/index.js]:
```
return provider ? provider.mainnetAddress : undefined
```

File: src/routes/scoreboard/components/RewardClaim.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+const RewardClaim = ({ mainnetAddress, openSetMainnetAddressModal }) => {
+  const showLoading = mainnetAddress === undefined
+  const hasRegistered = parseInt(mainnetAddress, 16) > 0
+
+  return (
+    <Block className={cx('rewardClaim')}>
+      <IconSection hasRegistered={hasRegistered} />
+      {showLoading ? (
```
Comment (apanizo) [inline src/routes/scoreboard/components/RewardClaim.js]: Ideally, when we have those kinds of conditions, components should try to avoid to be "Spaguetti". In my opinion, they are easy to understand. Something like:
```
....
{showLoading 
    ?  <LoadingIndicator />
    :  <RewardClaimAddress registered={ hasRegistered } /> 
}
....

----------------

const RewarClaimAddress = ({registered}) => (
  <Block>
      ...
  </Block>
)
```

Don't you think?

File: src/routes/scoreboard/components/RewardClaim.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+import Block from 'components/layout/Block'
+import Paragraph from 'components/layout/Paragraph'
+import Title from 'components/layout/Title'
+
+import * as css from './index.css'
+
+const cx = classNames.bind(css)
+
+const IconSection = ({ hasRegistered }) => (
+  <Block className={cx('iconSection')}>
```
Comment (apanizo) [inline src/routes/scoreboard/components/RewardClaim.js]:
```
<Block className={cx('iconSection')}>
   <i className="`icon icon--${hasRegistered ? 'ok' : 'warning'}`
</Block>
```
That would work?

File: src/components/SetMainnetAddress/index.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+const SetMainnetAddress = ({
+  submitting, handleSubmit, error, submitFailed, closeModal,
+}) => (
+  <Block className="setMainnetAddress">
+    <Block className="setMainnetAddressModal">
+      <a className="setMainnetAddressModal__close" href="javascript:void(0);" onClick={closeModal} />
+      <form onSubmit={handleSubmit}>
+        <Title className="setMainnetAddressModal__header">Setup claim Address</Title>
+        <Paragraph className="setMainnetAddressModal__disclaimer">Please register your Metamask address, where we can send your winning GNO tokens in case you finished in the top 100. Please note that you can register your address <em>only once</em>.</Paragraph>
```
Comment (apanizo) [inline src/components/SetMainnetAddress/index.js]: You use a Layout component and then in CSS you modify its core properties (Paragraph --> fontSize).

Ideally, the core props of a layout components should be specified as a regular Reeact props:
```
<Paragraph size="sm" />  //or something like that.
```

Also in this component you have used <p> and <Paragraph>. I would love to see this component converted using layout components and CSS in modules, but if you do not have time, I think is better if you remove completely layout components from here (just for consistency).

File: src/components/HeaderRewardStatus/index.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    if (hideRewardStatusInfo) {
+      return null
+    }
+
+    return (
+      <Block className={cx('headerReward')}>
+        <Block className="container">
+          <Block className={cx('headerRewardStatus')}>
+            <i className="icon icon--warning" />
+
```
Comment (apanizo) [inline src/components/HeaderRewardStatus/index.js]: Empty line? Typo?

File: src/api/gnosis.js (hunk 5 above)
Comment (apanizo) [reply src/api/gnosis.js]: Sure. No problem!

**Comment by apanizo** [comment]
Good job. Now the PR looks much better and organised. There is more "job" to do (I had no time to review css for example), but we do not have time. 

Approved,  but a couple of things to consider:

- Ideally, PRs should be smaller. For example, if you have added two widgets (one at the top of the window and other inside scoreboard view), one PR per widget should be better (faster to code, review and merge)
- I think components with a specific purpose (HeaderRewardStatus for example) should be private components of the parent's one (Header), even when that component has its one container (same logic as usual: making bigger and bigger **scr/components** makes that folder difficult to manage and understand, also when you want to check if you can reuse an existant widget).
- Another benefit of using layout components is... you need less css, see the example of ```SetMainnetAddress/index.js``` for some reason with layout components, the code would be simpler, smaller and easier to read, but do not worry, let's go step by step.
- Another small thing just for consistency, ideally commits should contain the ticket number :)
- We have to work together for making as a rule to include some storybook and JEST tests.

Thanks for the effort addressing all my comments.
